services/wfm/src/main/java/org/openkilda/wfm/topology/flow/service/FlowResourcesManager.java

```diff
@@ -0,0 +1,136 @@
+
+import lombok.ToString;
+import lombok.extern.slf4j.Slf4j;
+
+import java.util.Map;
+import java.util.Set;
+
+@Slf4j
+@ToString
+public class FlowResourcesManager {
+    private final ResourceCache resourceCache;
+
+    public FlowResourcesManager(ResourceCache resourceCache) {
+        this.resourceCache = resourceCache;
+    }
+
+    /**
+     * Clears the inner resource pools.
+     */
+    public void clear() {
+        resourceCache.clear();
+    }
+
+    /**
+     * Register the flow resources as used.
+     *
+     * @param flowPair the flow pair to register.
+     */
+    public void registerUsedByFlow(FlowPair flowPair) {
+        Flow forward = flowPair.getForward();
+        resourceCache.allocateCookie((int) (ResourceCache.FLOW_COOKIE_VALUE_MASK & forward.getCookie()));
+        if (!forward.isOneSwitchFlow()) {
+            // Don't allocate if one switch .. it is zero
+            // .. and allocateVlanId *will* allocate if it is zero, which consumes a very
+            // .. limited resource unnecessarily
+            resourceCache.allocateVlanId(forward.getTransitVlan());
+        }
+        resourceCache.allocateMeterId(forward.getSrcSwitch().getSwitchId(), forward.getMeterId());
+
+        Flow reverse = flowPair.getReverse();
+        if (!reverse.isOneSwitchFlow()) {
+            // Don't allocate if one switch .. it is zero
+            // .. and allocateVlanId *will* allocate if it is zero, which consumes a very
+            // .. limited resource unnecessarily
+            resourceCache.allocateVlanId(reverse.getTransitVlan());
+        }
+        resourceCache.allocateMeterId(reverse.getSrcSwitch().getSwitchId(), reverse.getMeterId());
+    }
+
+    /**
+     * Allocate free resources for the flow.
+     *
+     * @param flowPair the flow pair for resource allocation.
+     * @return a new flow pair (copied) with allocated resources set.
+     */
+    public FlowPair allocateFlow(FlowPair flowPair) {
+        log.debug("Allocate flow resources for {}.", flowPair);
+
+        int flowCookie = resourceCache.allocateCookie();
+
+        Flow forward = allocateFlowResources(flowPair.getForward());
+        forward.setCookie(flowCookie | ResourceCache.FORWARD_FLOW_COOKIE_MASK);
+
+        Flow reverse = allocateFlowResources(flowPair.getReverse());
+        reverse.setCookie(flowCookie | ResourceCache.REVERSE_FLOW_COOKIE_MASK);
+
+        return FlowPair.builder().forward(forward).reverse(reverse).build();
+    }
+
+    private Flow allocateFlowResources(Flow flow) {
+        Flow.FlowBuilder newFlow = flow.toBuilder();
+        // Don't allocate a vlan for a single switch flow.
+        newFlow.transitVlan(!flow.isOneSwitchFlow() ? resourceCache.allocateVlanId() : 0);
+
+        if (flow.getBandwidth() > 0L) {
+            newFlow.meterId(resourceCache.allocateMeterId(flow.getSrcSwitch().getSwitchId()));
+        } else {
+            // When the flow is unmetered.
+            newFlow.ignoreBandwidth(true);
```

Comment:
Nice idea... but can become a problem. If flow will be update and bandwidth will be changed from 0 to some meaningful value, we will be unable to determine correct/original value of `ignore-bandwidth` flag... From other side, we receive whole flow object during update, so correct value for "ignore-bandwith" flag is caller responsibility.